src/main/java/io/confluent/kafkarest/ProducerPool.java

```diff
@@ -186,16 +199,57 @@ private static String getBootstrapBrokers(ZkUtils zkUtils) {
       String topic,
       Integer partition,
       EmbeddedFormat recordFormat,
+      SecurityContext securityContext,
       SchemaHolder schemaHolder,
       Collection<? extends ProduceRecord<K, V>> records,
       ProduceRequestCallback callback
   ) {
     ProduceTask task = new ProduceTask(schemaHolder, records.size(), callback);
     log.trace("Starting produce task " + task.toString());
-    RestProducer restProducer = producers.get(recordFormat);
+    RestProducer restProducer = getRestProducer(recordFormat, getJaasModule(securityContext));
     restProducer.produce(task, topic, partition, records);
   }
 
+  private RestProducer getRestProducer(EmbeddedFormat recordFormat, JaasModule jaasModule) {
+    RestProducer restProducer;
+    if (jaasModule != null) {
+      if (!authenticatedProducers.containsKey(jaasModule)) {
+        restProducer = buildAuthenticatedProducer(recordFormat, jaasModule);
+        authenticatedProducers.put(jaasModule, restProducer);
```

Comment:
each user principal will have its own producer, if we have many principals, like in our case we have 20/25 principals, we will not get a good throughput from the producer if they all producing same topic. If the max.buffer.size is high enough we could also run into memory issues with many producers.